p2p/proto/consensus.proto

```diff
@@ -1,42 +1,80 @@
-import "p2p/proto/state.proto";
+import "common.proto";
+import "transaction.proto";
+import "block.proto";
 import "google/protobuf/timestamp.proto";
 
 
 // WIP - will change
 
-message Proposal {
-    uint64                    block_number      = 2;
-    uint32                    round             = 3;
-    uint32                    pol               = 4;  // proof of lock
-    Hash                      block_header_hash = 5;
-    google.protobuf.Timestamp timestamp         = 6;
-    ConsensusSignature        signature         = 7;
+
+message BlockLocation {
+    Hash   parent_hash = 1;
+    uint64 height      = 2;
+    uint64 fork_id     = 3;
 }
 
-// A block proposal is a series of (Transactions+, StateDiff)* BlockHeader
+enum Step {
+    Propose = 1;
+    Prevote = 2;
+    Precommit = 3;
+}
 
 message Vote {
-    enum  Type {
-        UNKNOWN   = 0;
-        Proposal  = 1;
-        Prevote   = 2;
-        Precommit = 3;
+    enum  VoteType {
+        Prevote   = 1;
+        Precommit = 2;
     };
 
-    Proposal           proposal          = 1;
-    Address            validator_address = 2;
-    int32              validator_index   = 3;  // ???
-    ConsensusSignature signature         = 4;
+    // We use a type field to distinguish between prevotes and precommits instead of different
+    // messages, to make sure the data, and therefore the signatures, are unambiguous between
+    // Prevote and Precommit.
+    VoteType           vote_type      = 3;
+    BlockLocation      block_location = 4;
+    uint32             round          = 5;
+    // This is optional since a vote can be NIL.
+    optional Hash      block_hash     = 6;
+    // Identifies the voter.
+    Address            voter          = 7;
+    // Voter's signature of the above fields.
+    ConsensusSignature signature      = 8;
+}
+
+message ProposalInit {
+    BlockLocation             block_location = 1;
+    uint32                    proposal_round = 2;
+    Address                   proposer       = 4;
 }
 
-message CreateBlock {
+// Produce the full Tendermint Proposal message once the proposer node has finished building it.
+// Tendermint just ID's the proposal's content, block_hash.
+message ProposalFin {
+    BlockLocation      block_location = 1;
+    uint32             proposal_round = 2;
+    optional uint32    valid_round    = 3;
+    Address            proposer       = 4;
+    // Validators must verify this matches the block_hash they calculated from the content.
+    Hash               block_hash     = 5;
+}
+
+// The timestamp of a proposal can impact consensus, specifically the lower bound applied. If nodes
+// apply a lower bound validation based on their local time, then we risk a scenario where in round
+// `R` proposal P is locked. Then in a later round the timestamp in P has gone stale. Therefore the
+// lower bound should be "greater than the previous timestamp". Upper bounds don't suffer from this
+// problem.
+message Proposal {
```

Comment:
We also need to pass Vote messages on the same topic.

Reply:
I created the ConsensusMessage proto below.